### PR TITLE
feat(stream): add rate limit for dml

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support DML rate limit. This is done by:
1. Splitting the DML chunks inside `into_stream_with_rate_limit`, to ensure rate limiting can happen for chunks larger than the rate limit.
2. Updating the DMLExecutor's stream to bias on barrier, to allow barriers to bypass the DML side.
3. Adding a `FlowControlExecutor` after the DML executor, to rate limit the chunks.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

</details>

Rate limit now impacts DML statements also.
```
SET STREAMING_RATE_LIMIT=1;
CREATE TABLE t(v1 int);
```

The DML to `t` will be rate limited to 1 record / s

### CODE REVIEW

File: src/stream/src/executor/dml.rs [hunk 1]
```diff
@@ -80,7 +84,13 @@ impl DmlExecutor {
         table_version_id: TableVersionId,
         column_descs: Vec<ColumnDesc>,
         chunk_size: usize,
+        rate_limit: Option<NonZeroU64>,
     ) -> Self {
+        // If chunk_size > rate_limit, we should decrease its size, so rate limit can take effect.
+        let chunk_size = rate_limit
+            .map(|rate_limit| std::cmp::min(chunk_size, rate_limit.get() as usize))
+            .unwrap_or(chunk_size);
+        tracing::debug!("chunk_size: {}", chunk_size);
```
Comment (kwannoel) [inline src/stream/src/executor/dml.rs:93]:
```suggestion
```

File: src/stream/src/executor/dml.rs [hunk 2]
```diff
@@ -158,11 +175,13 @@ impl DmlExecutor {
                             for chunk in vec {
                                 for (op, row) in chunk.rows() {
                                     if let Some(chunk) = builder.append_row(op, row) {
+                                        tracing::trace!("flush batch_group on barrier");
```
Comment (kwannoel) [inline src/stream/src/executor/dml.rs:178]:
```suggestion
```

File: src/stream/src/executor/dml.rs [hunk 3]
```diff
@@ -158,11 +175,13 @@ impl DmlExecutor {
...
                                         yield Message::Chunk(chunk);
                                     }
                                 }
                             }
                             if let Some(chunk) = builder.take() {
+                                tracing::trace!("flush batch_group on barrier");
```
Comment (kwannoel) [inline src/stream/src/executor/dml.rs:184]:
```suggestion
```

File: src/stream/src/executor/dml.rs [hunk 4]
```diff
@@ -198,36 +217,42 @@ impl DmlExecutor {
                                     for chunk in vec {
                                         for (op, row) in chunk.rows() {
                                             if let Some(chunk) = builder.append_row(op, row) {
+                                                tracing::trace!("yielding, txn buffer large");
```
Comment (kwannoel) [inline src/stream/src/executor/dml.rs:220]:
```suggestion
```

File: src/stream/src/executor/dml.rs [hunk 5]
```diff
@@ -198,36 +217,42 @@ impl DmlExecutor {
...
                                                 yield Message::Chunk(chunk);
                                             }
                                         }
                                     }
                                     if let Some(chunk) = builder.take() {
+                                        tracing::trace!("yielding, txn buffer large");
```
Comment (kwannoel) [inline src/stream/src/executor/dml.rs:226]:
```suggestion
```

**Comment by kwannoel** [comment]
Testing it in ci

**Comment by kwannoel** [comment]
The approach taken here can starve the DML side when table has a source attached.

We should rate limit the dml using permits instead. Updating the implementation.

Thanks @chenzl25 for mentiiioning it.

**Comment by kwannoel** [comment]
Currently all DML run in batch are "transactional", meaning we insert a `TxnBegin` and `TxnEnd` for it. Paused for now since I'm not sure how to solve this case.

Also another refactor is needed for this PR, to move the rate limit logic from stream side to batch side. This is because stream side cannot bias to the barrier side, otherwise it will starve the dml side when there's table with source.

**Comment by chenzl25** [comment]
> Also another refactor is needed for this PR, to move the rate limit logic from stream side to batch side. This is because stream side cannot bias to the barrier side, otherwise it will starve the dml side when there's table with source.

FYI. The new version of the streaming graph of the table doesn't need to bias to the barrier side, because the new version uses a `UNION ALL` to combine the source input and the dml input.

```sql
 StreamMaterialize { columns: [o_orderkey, o_custkey, o_orderstatus, o_totalprice, o_orderdate, o_orderpriority, o_clerk, o_shippriority, o_comment], stream_key: [o_orderkey], pk_columns: [o_orderkey], pk_conflict: Overwrite }
 └─StreamUnion { all: true }
   ├─StreamExchange { dist: HashShard(o_orderkey) }
   │ └─StreamSource { source: orders, columns: [o_orderkey, o_custkey, o_orderstatus, o_totalprice, o_orderdate, o_orderpriority, o_clerk, o_shippriority, o_comment] }
   └─StreamExchange { dist: HashShard(o_orderkey) }
     └─StreamDml { columns: [o_orderkey, o_custkey, o_orderstatus, o_totalprice, o_orderdate, o_orderpriority, o_clerk, o_shippriority, o_comment] }
       └─StreamSource
```

**Pull request closed without merging**